witness/src/Dscp/Witness/Workers/Worker.hs

```diff
@@ -6,72 +6,88 @@ module Dscp.Witness.Workers.Worker
-                      "with proof" +|| proof ||+ ", propagating"
 
-----------------------------------------------------------------------------
--- Ping/pong workers
-----------------------------------------------------------------------------
+    -- We ask for a tip on startup to synchronise.
+    bootstrap :: ClientEnv NetTag -> m ()
+    bootstrap btq = do
+        logDebug "Bootstrapped, asking for a tip"
+        cliSend btq Nothing GetTipMsg
 
-_witnessTxWorker :: forall ctx m. WitnessWorkMode ctx m => Worker m
-_witnessTxWorker =
-    Worker "txWorker" [msgType @PongTx] [] (\btq -> action btq `catchAny` handler)
-  where
-    handler e = logError $ fromString $ "Exception in txWorker: " <> show e
-    action :: ClientEnv ZmqTcp -> m ()
+    action :: ClientEnv NetTag -> m ()
     action btq = do
-      logInfo "Started witness tx worker"
-      forever $ do
-        cliSend btq Nothing PingTx
-        (nId,PongTx txt) <- cliRecvResp btq (-1)
-        logInfo $ "Heard pongtx: " +|| txt ||+ " from " +|| nId ||+ ""
-        liftIO $ threadDelay 1000000
+        logDebug "Started block update worker"
 
-_witnessBlkWorker :: forall ctx m. WitnessWorkMode ctx m => Worker m
-_witnessBlkWorker =
-    Worker "blkWorker" [msgType @PongBlk] [] (\btq -> action btq `catchAny` handler)
-  where
-    handler e = logError $ fromString $ "Exception in txWorker: " <> show e
-    action :: ClientEnv ZmqTcp -> m ()
-    action btq = do
-      liftIO $ threadDelay 500000 -- for clarity of wor
-      logInfo "Started witness blk worker"
-      forever $ do
-        cliSend btq Nothing PingBlk
-        (nId,PongBlk txt) <- cliRecvResp btq (-1)
-        logInfo $ "Heard pongblk: " +|| txt ||+ " from " +|| nId ||+ ""
-        liftIO $ threadDelay 1000000
+        let processPubBlock nId (PubBlock block) = do
+                logInfo $ "New block was published: " +| hashF (headerHash block) |+ ""
+                processNewBlock nId btq block
+        let processTipMsg nId (TipMsg block) = do
+                logInfo $ "Got network tip: " +| hashF (headerHash block) |+ ""
+                processNewBlock nId btq block
+        let processBlocks _nId (NoBlocksMsg e) = do
+                logWarning $ "Processing blocks: got none: " +|| e ||+ ""
+            processBlocks _nId (BlocksMsg blocks) = do
+                logDebug $ "Processing blocks"
+                applyManyBlocks blocks
+
+        forever $ do
+            logDebug "blockUpdateWorker: receiving"
+            cliRecv btq (-1) [ ccallback processPubBlock
+                             , ccallback processTipMsg
+                             , ccallback processBlocks
+                             ]
+
+    processNewBlock nId btq block = do
+        let header = bHeader block
+        tip <- runSdM getTipHeader
+        let tipHash = headerHash tip
+        if | hPrevHash header == tipHash -> applyNewBlock block
+           | hDifficulty header > hDifficulty tip -> do
```

Comment:
Is the "difficulty" the length of the chain?

Reply:
Yes, it's a common notation.